### PR TITLE
Support kCTParagraphStyleSpecifierLineHeightMultiple

### LINKED ISSUE #1090: Support kCTParagraphStyleSpecifierLineHeightMultiple

### PR DESCRIPTION
**Pull request by aballway**
Adds support for kCTParagraphStyleSpecifierLineHeightMultiple.  DWrite only allows line height multiples if the line height is kept constant through the frame, which is not guaranteed, so we need to manually manipulate the line origins.  Also removes the redundant line origin variable in _CTLine.

Fixes #1090

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1298)

### CODE REVIEW

File: Frameworks/CoreText/CTFrame.mm [hunk 1]
```diff
@@ -61,8 +61,9 @@ CFRange CTFrameGetVisibleStringRange(CTFrameRef frame) {
     _CTFrame* framePtr = static_cast<_CTFrame*>(frame);
     CFIndex count = 0;
     if (framePtr) {
-        for (_CTLine* line in static_cast<id<NSFastEnumeration>>(framePtr->_lines)) {
-            if (line->_lineOrigin.y < framePtr->_frameRect.size.height && line->_lineOrigin.y > 0) {
+        for (size_t i = 0; i < framePtr->_lineOrigins.size(); ++i) {
+            if (framePtr->_lineOrigins[i].y < framePtr->_frameRect.size.height && framePtr->_lineOrigins[i].y > 0) {
```
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTFrame.mm]: > framePtr->_lineOrigins[i].y > 0 [](start = 81, length = 32)

switch around and move this check first.

File: Frameworks/CoreText/CTFramesetter.mm [hunk 2]
```diff
@@ -27,15 +27,48 @@ @implementation _CTFramesetter : NSObject
...
+                    totalShifted += lineHeightMultiple * (ret->_lineOrigins[i].y - lastOriginY);
+                    lastOriginY = ret->_lineOrigins[i].y;
+                    ret->_lineOrigins[i].y += totalShifted;
+                }
+
+                // Adjust framesize to account for changes in lineheights
+                ret->_frameRect.size.height += totalShifted;
+            }
+        }
+    }
```
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTFramesetter.mm]: Although you don't "prefer" early returns. This code would be so much simpler and much simpler to read if you used early returns.
We have if layers that aren't the greatest looking for code readability.

File: Frameworks/CoreText/CTFramesetter.mm [hunk 3]
```diff
@@ -27,15 +27,48 @@ @implementation _CTFramesetter : NSObject
...
+        if (settings) {
+            // DWrite only gives manual control of lineheight when it is constant through a frame
+            // We need to shift each line by the difference in lineheight manually
+            CGFloat lineHeightMultiple = 0.0f;
+            if (CTParagraphStyleGetValueForSpecifier(settings,
+                                                     kCTParagraphStyleSpecifierLineHeightMultiple,
+                                                     sizeof(lineHeightMultiple),
+                                                     &lineHeightMultiple) &&
+                lineHeightMultiple > 0) {
+                // The actual ratio we need to change the line height by is multiple - 1
```
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTFramesetter.mm]: >multiple [](start = 76, length = 8)

lineHeightMultiple

File: tests/unittests/CoreText/CTParagraphStyleTests.mm [hunk 4]
```diff
@@ -119,4 +119,59 @@
...
     EXPECT_NEAR(thirdAdvances[2].width, firstAdvances[2].width - c_diff, c_errorDelta);
     EXPECT_NEAR(thirdAdvances[3].width, firstAdvances[3].width - c_diff, c_errorDelta);
+}
+
+TEST(CoreText, LineSpaceMultiple) {
+    const static float c_errorDelta = 0.001f;
+    CGPathRef path = CGPathCreateWithRect(CGRectMake(0, 0, FLT_MAX, FLT_MAX), nullptr);
+
+    NSMutableAttributedString* normalString = [[[NSMutableAttributedString alloc] initWithString:@"TEST\nTEST\nTEST"] autorelease];
+    CTFramesetterRef normalFramesetter = CTFramesetterCreateWithAttributedString((CFAttributedStringRef)normalString);
```
Comment (msft-Jeyaram) [inline tests/unittests/CoreText/CTParagraphStyleTests.mm]: >CFAttributedStringRef [](start = 82, length = 21)

proper casting?

File: tests/unittests/CoreText/CTParagraphStyleTests.mm [hunk 5]
```diff
@@ -119,4 +119,59 @@
...
+    NSMutableAttributedString* doubleString = [[[NSMutableAttributedString alloc] initWithString:@"TEST\nTEST\nTEST"] autorelease];
+    CGFloat twice = 2.0f;
+    CTParagraphStyleSetting settings[1] = {
+        {.spec = kCTParagraphStyleSpecifierLineHeightMultiple, .valueSize = sizeof(CGFloat), .value = &twice }
+    };
+    [doubleString setAttributes:@{
+        static_cast<NSString*>(kCTParagraphStyleAttributeName) : (id)CTParagraphStyleCreate(settings, 1)
+    }
+                          range:NSMakeRange(0, 14)];
+
```
Comment (msft-Jeyaram) [inline tests/unittests/CoreText/CTParagraphStyleTests.mm]: bad formatting?

File: tests/unittests/CoreText/CTParagraphStyleTests.mm [hunk 6]
```diff
@@ -119,4 +119,59 @@
...
+                        range:NSMakeRange(0, 14)];
+
+    CTFramesetterRef halfFramesetter = CTFramesetterCreateWithAttributedString((CFAttributedStringRef)halfString);
+    CFAutorelease(halfFramesetter);
+    CTFrameRef halfFrame = CTFramesetterCreateFrame(halfFramesetter, {}, path, nullptr);
+    CFAutorelease(halfFrame);
+    CGPoint halfOrigins[3];
+    CTFrameGetLineOrigins(halfFrame, {}, halfOrigins);
+    EXPECT_NEAR(half * normalLineSpaceFirst, halfOrigins[1].y - halfOrigins[0].y, c_errorDelta);
+    EXPECT_NEAR(half * normalLineSpaceSecond, halfOrigins[2].y - halfOrigins[0].y, c_errorDelta);
```
Comment (msft-Jeyaram) [inline tests/unittests/CoreText/CTParagraphStyleTests.mm]: this is a duplicate of the above with slightly different configuration. Refactor the code so you can use it generically and also can add more tests easily.

**Comment by msft-Jeyaram** [comment]
:clock1:

File: Frameworks/CoreText/CTFrame.mm (hunk 1 above)
Comment (aballway) [inline Frameworks/CoreText/CTFrame.mm]: whoops this check isn't actually needed any longer

File: Frameworks/CoreText/CTFramesetter.mm (hunk 2 above)
Comment (aballway) [inline Frameworks/CoreText/CTFramesetter.mm]: I hate to admit it but you're right in this case, too many layers

File: Frameworks/CoreText/CTFramesetter.mm [hunk 7]
```diff
@@ -27,15 +27,48 @@ @implementation _CTFramesetter : NSObject
...
     // Call _DWriteWrapper to get _CTLine object list that makes up this frame
     _CTTypesetter* typesetter = static_cast<_CTTypesetter*>(framesetter->_typesetter);
     if (range.length == 0L) {
         range.length = typesetter->_characters.size();
     }
 
-    return [_DWriteGetFrame(static_cast<CFAttributedStringRef>(typesetter->_attributedString.get()), range, frameSize) retain];
+    _CTFrame* ret = _DWriteGetFrame(static_cast<CFAttributedStringRef>(typesetter->_attributedString.get()), range, frameSize);
+
+    // Trying to access attributes without any text with throw an error
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/CTFramesetter.mm]: nit: typo: // Trying to access attributes without any text ~~with~~ _will_ throw an error

File: tests/unittests/CoreText/CTParagraphStyleTests.mm [hunk 8]
```diff
@@ -119,4 +120,52 @@
...
+
+public:
+    CGFloat normalLineSpaceFirst = 0;
+    CGFloat normalLineSpaceSecond = 0;
+    CGPathRef path;
+};
+
+TEST_P(CoreTextLineSpaceMultipleTest, OriginsShouldBeMovedByRatio) {
+    NSMutableAttributedString* multipleString = [[[NSMutableAttributedString alloc] initWithString:@"TEST\nTEST\nTEST"] autorelease];
+    CGFloat param = GetParam();
```
Comment (ms-jihua) [inline tests/unittests/CoreText/CTParagraphStyleTests.mm]: >param [](start = 12, length = 5)

please give this a more meaningful name

File: Frameworks/CoreText/CTFramesetter.mm [hunk 9]
```diff
@@ -27,15 +27,51 @@ @implementation _CTFramesetter : NSObject
...
+
+    // Trying to access attributes without any text will throw an error
+    if (range.length <= 0L) {
+        return [ret retain];
+    }
+
+    CTParagraphStyleRef settings = static_cast<CTParagraphStyleRef>(
+        [typesetter->_attributedString attribute:static_cast<NSString*>(kCTParagraphStyleAttributeName) atIndex:0 effectiveRange:nullptr]);
+
+    if (settings == nil) {
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTFramesetter.mm]: >nil [](start = 20, length = 3)

nit: nullptr, since it's not objc

File: Frameworks/CoreText/CTFramesetter.mm [hunk 10]
```diff
@@ -27,15 +27,51 @@ @implementation _CTFramesetter : NSObject
...
+    RETURN_NULL_IF(framesetter == nil);
 
     // Call _DWriteWrapper to get _CTLine object list that makes up this frame
     _CTTypesetter* typesetter = static_cast<_CTTypesetter*>(framesetter->_typesetter);
     if (range.length == 0L) {
         range.length = typesetter->_characters.size();
     }
 
-    return [_DWriteGetFrame(static_cast<CFAttributedStringRef>(typesetter->_attributedString.get()), range, frameSize) retain];
+    _CTFrame* ret = _DWriteGetFrame(static_cast<CFAttributedStringRef>(typesetter->_attributedString.get()), range, frameSize);
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTFramesetter.mm]: >_CTFrame [](start = 4, length = 8)

If you do a `StrongId<_CTFrame>` and then `.detach()` when you return, this will memory manage sanely.

**Comment by DHowett-MSFT** [comment]
:shipit:

**Comment by ms-jihua** [comment]
:shipit:
